packages/eslint-plugin/src/rules/no-shadow.ts

```diff
@@ -63,7 +70,7 @@ export default createRule<Options, MessageIds>({
             type: 'string',
             description:
               'Whether to report shadowing before outer functions or variables are defined.',
-            enum: ['all', 'functions', 'never'],
+            enum: ['all', 'functions', 'functions-and-types', 'never', 'types'],
```

Comment:
> I've expanded the existing `hoist` configuration to include `types` and `functions-and-types`. I don't think this is optional, as permutations need to be manually set ...

Yeah, I'm thinking this exposes a general learning: don't use `enum` option types for _"when do I X?"_ questions. Eventually folks will want to do something that could be better expressed as an object. I.e. here:

```ts
{ optionName: { functions: true, types: true, variables: true } }
```

...but, to your point, I think it would be even more disruptive to extend the base rule option to a wholly different format in that way. For now I think the `enum` is fine. If this rule gets even more complicated over time then we can always switch to the object later.

tl;dr: agreed as-is 👍